### PR TITLE
Feature/stereo

### PR DESCRIPTION
**Pull request by ZacharyTaylor**
Big change to the image undistort node.
-Switched to a new class for storing and modifying camera parameters, as using camera_info messages was fairly ugly and quite limiting.
-Added the ability to get output parameters from a ros topic or have them automatically set to good values.
-Added the ability to do rectifications
-Added the ability to convert images and scale them easily
-Added a second node stereo_info_node that sets up all of the camera info needed to do dense stereo reconstruction.
-Added a launchfile that uses stereo_info_node and 2 undistort nodes in combination with stereo_image_proc to generate stereo depthmaps.

### CODE REVIEW

**Comment by ZacharyTaylor** [comment]
Switched to nodelets running inside nodes. This allows the creation of a new dense_stereo_node that performs the full stereo pipeline inside a single node. @helenol @raghavkhanna

File: include/image_undistort/undistorter.h [hunk 1]
```diff
@@ -36,8 +36,8 @@ class Undistorter {
  private:
   const CameraParametersPair used_camera_parameters_pair_;
 
-  cv::Mat map_x_;
-  cv::Mat map_y_;
+  cv::UMat map_x_;
+  cv::UMat map_y_;
 };
 }
```
Comment (raghavkhanna) [inline include/image_undistort/undistorter.h]: I think this is OpenCV3 which may be the default in 16.04, however, can we keep the Mat version in a different branch for the people still on 14.04?

**Comment by raghavkhanna** [comment]
Works great @ZacharyTaylor :).
Here are some screenshots of this in action. Works well on grass as well. Uses between 30 -60% of one CPU over 3 threads on the NUC i7 (without GPU processing). Thanks a lot.
screenshot from 2016-11-17 00 58 46
screenshot from 2016-11-17 00 59 37

**Comment by ZacharyTaylor** [comment]
@raghavkhanna I added some flags so hopefully it compiles as is under opencv 2 without issue now. I also changed the optimization flag from -O3 to -Ofast which made a big difference on my machine (speedup of 20%+)

File: src/camera_parameters.cpp [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+  // The general case is even more of a nightmare, so we constrain the problem
+  // such that the center of focus must be in the center of the final image.
+
+  // as we are missing the forward projection model we iteratively estimate
+  // image size assuming a linear relationship between warping and size at
+  // each
+  // step
+  for (size_t i = 0; i < kFocalLengthEstimationAttempts; ++i) {
+    // get list of edge points to check
+    std::vector<Eigen::Vector2d, Eigen::aligned_allocator<Eigen::Vector2d>>
```
Comment (ZacharyTaylor) [inline src/camera_parameters.cpp]: The undistorter needs these values as doubles as it makes no assumption about points coming from a 2d grid.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,30 +1,91 @@
 cmake_minimum_required(VERSION 2.8.3)
+
 project(image_undistort)
 
-find_package(catkin_simple REQUIRED)
```
Comment (ZacharyTaylor) [inline CMakeLists.txt]: I followed the realsense driver as an example when setting up building nodelets which didn't use catkin simple and didn't switch back after I got things running

**Pull request merged**